tests/unit/test_http_session.py

```diff
@@ -0,0 +1,196 @@
+        headers = self.proxy_config.proxy_headers_for(self.url)
+        self.assertEqual({}, headers)
+
+    def test_proxy_for_url_no_slashes(self):
+        self.update_http_proxy('localhost:8081/')
+        proxy_url = self.proxy_config.proxy_url_for(self.url)
+        self.assertEqual('http://localhost:8081/', proxy_url)
+
+    def test_proxy_for_url_no_protocol(self):
+        self.update_http_proxy('//localhost:8081/')
+        proxy_url = self.proxy_config.proxy_url_for(self.url)
+        self.assertEqual('http://localhost:8081/', proxy_url)
+
+    def test_fix_proxy_url_has_protocol_http(self):
+        proxy_url = self.proxy_config.proxy_url_for(self.url)
+        self.assertEqual('http://localhost:8081/', proxy_url)
+
+
+class TestHttpSessionUtils(unittest.TestCase):
+    def test_get_cert_path_path(self):
+        path = '/some/path'
+        cert_path = get_cert_path(path)
+        self.assertEqual(path, cert_path)
+
+    def test_get_cert_path_certifi_or_default(self):
+        with patch('botocore.httpsession.where') as where:
+            path = '/bundle/path'
+            where.return_value = path
+            cert_path = get_cert_path(True)
+            self.assertEqual(path, cert_path)
+
+
+class TestURLLib3Session(unittest.TestCase):
+    def setUp(self):
+        self.request = AWSRequest(
+            method='GET',
+            url='http://example.com/',
+            headers={},
+            data=b'',
+        )
+
+        self.response = Mock()
+        self.response.headers = {}
+        self.response.stream.return_value = b''
+
+        self.pool_manager = Mock()
+        self.connection = Mock()
+        self.connection.urlopen.return_value = self.response
+        self.pool_manager.connection_from_url.return_value = self.connection
+
+        self.pool_patch = patch('botocore.httpsession.PoolManager')
+        self.proxy_patch = patch('botocore.httpsession.proxy_from_url')
+        self.pool_manager_cls = self.pool_patch.start()
+        self.proxy_manager_fun = self.proxy_patch.start()
+        self.pool_manager_cls.return_value = self.pool_manager
+        self.proxy_manager_fun.return_value = self.pool_manager
+
+    def tearDown(self):
+        self.pool_patch.stop()
+
+    def assert_request_sent(self, headers=None, body=None, url='/'):
+        if headers is None:
+            headers = {}
+
+        self.connection.urlopen.assert_called_once_with(
+            method=self.request.method,
+            url=url,
+            body=body,
+            headers=headers,
+            retries=False,
+            assert_same_host=False,
+            preload_content=False,
+            decode_content=False,
+            request_url=ANY,
+        )
+
+    def test_forwards_max_pool_size(self):
+        URLLib3Session(max_pool_connections=22)
+        self.pool_manager_cls.assert_called_with(
+            maxsize=22,
+            timeout=ANY,
+            strict=True,
+            ssl_context=ANY,
```

Comment:
We don't seem to be actually testing we're passing the correct ssl_context to the pool manager in any of our tests.  We could pass `None` and these tests would still pass.

I think we need at least one test that ensures it's a sane value.

Reply:
Strictly speaking, `None` is a sane value (urllib3 will default it for us). I'm not even sure I could assert the type of object passed either... I'll see what I can do.